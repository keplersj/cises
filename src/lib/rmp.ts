import queryString from "query-string";

type Professor = {
  city_state_s: string;
  averageratingscore_rf: number;
  pk_id: number;
  schoolcity_s: string;
  schoolstate_full_s: string;
  pict_thumb_name_s: string;
  id: string;
  pageviews_i: number;
  averagehelpfulscore_rf: number;
  schoolcountry_s: string;
  schoolname_s: string;
  status_i: number;
  teachermiddlename_t: string;
  averagehotscore_rf: number;
  schoolstate_s: string;
  rated_date_dt: string;
  teacherfullname_s: string;
  teacherdepartment_s: string;
  total_number_of_ratings_i: number;
  visible_i: number;
  content_type_s: string;
  averageeasyscore_rf: number;
  schoolid_s: string;
  teacherfirstname_t: string;
  teacherlastname_t: string;
  averageclarityscore_rf: number;
  schoolwebpage_s: string;
  timestamp: string;
  tag_s_mv: string[];
  tag_id_s_mv: string[];
};

export async function getProfessorFromSchool(
  teacherName: string,
  schoolName: string
): Promise<Professor> {
  const query = queryString.stringify({
    bf: "pow(total_number_of_ratings_i,2.1)",
    callback: "noCB",
    defType: "edismax",
    fq: `schoolname_t:"${schoolName}"`,
    group: "off",
    "group.field": "content_type_s",
    "group.limit": "20",
    "json.wrf": "noCB",
    q: teacherName,
    qf:
      "teacherfirstname_t^2000 teacherlastname_t^2000 teacherfullname_t^2000 teacherfullname_autosuggest",
    rows: ["20", "20"],
    siteName: "rmp",
    solrformat: "true",
    sort: "score desc",
    wt: "json"
  });

  const res = await fetch(
    `https://search-production.ratemyprofessors.com/solr/rmp/select/?${query}`
  );

  function noCB(obj: any) {
    return obj;
  }

  const { response } = eval(await res.text());

  return response.docs[0] as Professor;
}
