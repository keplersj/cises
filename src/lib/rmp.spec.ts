import * as rmp from "./rmp";

describe("Rate My Professors", () => {
  describe("getProfessorFromSchool", () => {
    beforeEach(() => {
      fetchMock.resetMocks();
    });

    it("provides object from RMP's callback function", async () => {
      fetchMock.mockResponseOnce("noCB({ response: { docs: [{}] } })");

      const response = await rmp.getProfessorFromSchool(
        "Teacher Name",
        "School Name"
      );

      expect(response).toEqual({});
    });

    it("uses fetch to get data from the Rate My Professors backend", async () => {
      let query = "";

      fetchMock.mockImplementationOnce((fetchQuery: string) => {
        query = fetchQuery;

        fetchMock.mockResponseOnce("noCB({ response: { docs: [{}] } })");

        return fetchMock(fetchQuery);
      });

      await rmp.getProfessorFromSchool("Teacher Name", "School Name");

      expect(fetchMock).toHaveBeenCalled();
      expect(query).not.toEqual("");
      expect(query).toMatchSnapshot();
    });
  });
});
