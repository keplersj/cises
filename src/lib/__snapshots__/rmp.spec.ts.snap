// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rate My Professors getProfessorFromSchool uses fetch to get data from the Rate My Professors backend 1`] = `"https://search-production.ratemyprofessors.com/solr/rmp/select/?bf=pow%28total_number_of_ratings_i%2C2.1%29&callback=noCB&defType=edismax&fq=schoolname_t%3A%22School%20Name%22&group=off&group.field=content_type_s&group.limit=20&json.wrf=noCB&q=Teacher%20Name&qf=teacherfirstname_t%5E2000%20teacherlastname_t%5E2000%20teacherfullname_t%5E2000%20teacherfullname_autosuggest&rows=20&rows=20&siteName=rmp&solrformat=true&sort=score%20desc&wt=json"`;
